pkg/destroy/kubevirt/destroyer.go

```diff
@@ -0,0 +1,179 @@
+package kubevirt
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/api/errors"
+	apilabels "k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	ickubevirt "github.com/openshift/installer/pkg/asset/installconfig/kubevirt"
+	"github.com/openshift/installer/pkg/destroy/providers"
+	"github.com/openshift/installer/pkg/types"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// deleteFunc is the interface a function needs to implement to be delete resources.
+type deleteFunc func(namespace string, listOpts metav1.ListOptions, kubevirtClient ickubevirt.Client) error
+
+// ClusterUninstaller holds the Metadata info needed to delete the tenantCluster resources from the infraCluster.
+type ClusterUninstaller struct {
+	Metadata types.ClusterMetadata
+	Logger   logrus.FieldLogger
+}
+
+// New returns KubeVirt Uninstaller from ClusterMetadata.
+func New(logger logrus.FieldLogger, metadata *types.ClusterMetadata) (providers.Destroyer, error) {
+	return &ClusterUninstaller{
+		Metadata: *metadata,
+		Logger:   logger,
+	}, nil
+}
+
+// Run is the entrypoint to start the uninstall process.
+func (uninstaller *ClusterUninstaller) Run() error {
+	namespace := uninstaller.Metadata.Kubevirt.Namespace
+
+	listOpts := metav1.ListOptions{LabelSelector: apilabels.FormatLabels(uninstaller.Metadata.Kubevirt.Labels)}
+	kubevirtClient, err := ickubevirt.NewClient()
+	if err != nil {
+		return err
+	}
+
+	deleteFuncs := []deleteFunc{uninstaller.deleteAllVMs, uninstaller.deleteAllDVs, uninstaller.deleteAllSecrets}
+	var results = make(chan error, len(deleteFuncs))
+	for i, del := range deleteFuncs {
+		go func(index int, delFunc deleteFunc) {
+			for {
+				err := delFunc(namespace, listOpts, kubevirtClient)
```

Comment:
Eventually, you are going to want to have one context that gets passed to all of your delete functions rather than having each delete function create its own contexts. This will become necessary once the migration to passing a context into `Run` for all the destroyers is completed. This is not something that you need to tackle for this PR, though.

Reply:
Changed all delete functions to use one context, created in the beginning of the Run function